### DISCUSSION TITLE
Selection / Inspection Policies

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#11952] [Internals and design (Q&A)]
cc @bokeh/dev 

* ref: https://github.com/bokeh/bokeh/pull/11899

## Summary

The above PR adds a new policy `NodesAndLinkedEdgesAndAdjacentNodes`. Under the current design, the extremely long name fits in perfectly, but its verboseness prompted me to think about some of the deficiencies of the current design that lead to awkward naming like that as the natural outcome.

More or less, the issue boils down to too much coupling. Specifically, it is not possible to specify policies for hits on nodes and edges separately. It's also not possible to parameterize the "distance" to include. These limitations prevent expressing different policies as compositions of simple building blocks, and mandates an entire new named Model any time any new combination is desired. 

I'd like to propose a new design that affords a flexible mechanism for composing these policies. First some motivating examples:

#### clicked nodes and nothing else
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    node_hit=[AdjacentNodes()]
)
```

#### nodes of clicked edges, and nothing else
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    edge_hit=[AdjacentNodes()]
)
```

#### clicked edges and nothing else
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    edge_hit=[AdjacentEdges()]
)
```

#### clicked edges and clicked nodes and nothing else
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    node_hit=[AdjacentNodes()],
    edge_hit=[AdjacentEdges()]
)
```

#### clicked nodes, their edges, and nodes 1 hop away
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    node_hit=[AdjacentNodes(hops=[0, 1]), AdjacentEdges()]
)
```

#### nodes that are 2 hops away from clicked nodes, and all edges leading to them
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    node_hit=[AdjacentNodes(hops=[2]), AdjacentEdges(hops=[1, 2])]
)
```

#### clicked nodes and custom edges and nodes away
```python
graph_renderer.selection_policy = GraphSelectionPolicy(
    node_hit=[AdjacentNodes(), CustomJSNodeSelection(...), CustomJSEdgeSelection(...)]
)
```

## Design Sketch
```python

class NodeSelectionPolicy:
    pass

class AdjacentNodes(NodeSelectionPolicy):
    """ Compute which nodes in a graph report selected/inspected after a hit test """

    hops = List(Int, default=[0], help="""
    What hop distances to include in selected results.

    For a node renderer hit-test, hop 0 nodes are the nodes that were clicked or tapped. 

    For an edge renderer hit-test, hop 0 nodes are the nodes that are on either end of an 
    edge that was clicked or tapped.  
    
    Hop 1 nodes are all the nodes nodes that are one edge traversal away from any hop 0 
    node (but not including  hop 0 nodes). Hop 2 nodes are all the nodes that are two edge 
    traversals away, etc. 
    """)

class EdgeSelectionPolicy:
    pass

class AdjacentEdges(EdgeSelectionPolicy):
    """ Compute which edges in a graph report selected/inspected after a hit test """

    hops = List(Int, default=[0], help="""
    What hop distances to include in selected results.

    For an edge renderer hit-test, hop 0 edges are the edges that were clicked or tapped. 

    For a node renderer hit-test, hop 0 edges are the edges that touch the node that was
    clicked or tapped.
    
    Hop 1 edges are all the nodes edges that are one node traversal away from any hop 0 
    edge (but not including  hop 0 edges). Hop 2 edges are all the nodes that are two node 
    traversals away, etc. 
    """)

class GraphSelectionPolicy:
    """ Specify how selections occur when nodes or edges are 'hit' """

    node_hit = List[Instance(NodeSelectionPolicy), help="""
    A series of node selection policies to apply when a node or edge is clicked or tapped. 
    Each policy is applied in order and the computed results of each are unioned together. 
    """)

    edge_hit = List[Instance(EdgeSelectionPolicy), help="""
    A series of edge selection policies to apply when a node or edge is clicked or tapped. 
    Each policy is applied in order and the computed results of each are unioned together. 
    """)
```

## Migration

Existing policies e.g. `NodesAndLinkedEdges` should become functions that return an appropriately configured `GraphSelectionPolicy` and also issues a deprecation warning. Suggest that any recent policies added only to `branch-3.0` before release simply be removed altogether. 

## Questions

Is there any reason to have a separate but very similar `GraphInspectionPolicy`? If these models simply compute the lists of indices, and don't actually perform selection or inspection (i.e. they are used by something else that actually does that step). Then I don't think any distinction is necessary. But in that case is there any better (more generic) terminology to indicate that the policies can apply equally to selections or inspections? 

Also/alternatively: should these rebrand as "hit testing policy" or something? There is already a separate, *different* notion of `Selection.selection_policy` that is for controlling how selections across different renderers are combined together, and that is completely unrelated to this.

What is the best way to expose "CustomJS" policies? The user will need to provide JS code for what to do when nodes are hit an/or when edges are hit.

How onerous will cycle detection be?

**Comment by g-parki** [comment]
This is nice! Thanks for preparing this @bryevdv 

> Is there any reason to have a separate but very similar `GraphInspectionPolicy`?

My vote's 👎 since they're so similar. I think the benefit of combining them outweighs the burden of maintaining both + trying to convey the nuance in the docs.

> any better (more generic) terminology to indicate that the policies can apply equally to selections or inspections?

I don't think `GraphSelectionPolicy` is bad. If I were a fresh user I wouldn't think twice about it.
At the expense of clashing with existing model names, `GraphHitTestPolicy` or `GraphHitTest` may be more neutral while getting the message across.
I considered `GraphInteractionPolicy`, but I think that's too generic to be useful.

**Reply by bryevdv** [reply]
> My vote's 👎 since they're so similar. I think the benefit of combining them outweighs the burden of maintaining both + trying to convey the nuance in the docs.

It's definitely also my preference, but I would also like to keep deprecated compatibility with the existing policies rather that drop them immediately, so it will take a little more actual investigation / experimentation to see if those are reconcilable goals (an if not, which to compromise). 

Re: naming, perhaps it would be instructive to consider the most verbose, descriptive name possible, to think about other synonyms or directions to trim. I guess truly this is the *"Policy for Extending Selections for Hit Tests on Graph Renderers"*. I think the thing that clicks there for me is that these are for taking an initial selection and extending/modifying them. So what about `ExtendedGraphSelection` ?

**Comment by bryevdv** [comment]
cc @p-himik would love your thoughts if you have time for a quick look

**Reply by p-himik** [reply]
Thanks for pinging.

Regarding the overall idea - as you have probably guessed, I definitely like the idea of being able to compose things together from tiny components. :) However, while your initial proposal does look good, the concrete node and edge selection policy classes make me think whether it could be refined further.

Is it a common scenario where a user wants to select/inspect nodes/edges based on plain topology and not on some properties specific to the particular graph data? "Highlight all edges of the selected node" does sound plausible, but "highlight twice removed edges" IMO sounds much less plausible than e.g. "highlight all edges that have `weight < 0.5`" or "highlight all nodes that are connected to the selected node with a path shorter than 10". My examples are definitely harder to implement than the ability to specify arbitrary `hops`, so they should probably be taken care of by some `CustomJS` or custom user subclasses. But if such scenarios are at least as useful as having `hops`, then I would avoid implementing `hops` altogether. No reason to provide a skewed support for graph theory only because we can. And if you think that we should support graphs on a more deeper level then I still feel that there might be a mechanism that would enable a uniform and flexible interface to have arbitrary filters without having to write too much boilerplate code. Let me know, and if so, I'll try to come up with something in a few days.


Regarding naming - I think it's a _very_ good idea to always keep the name of the package in mind. A class does not exist in a vacuum - it exists in its package, it belongs to it. If the package already has the word `graph` in it, there is no reason to include `Graph` in any class names (unless it's some abstract name or you want to oppose it to something that's not a graph but is conceptually on the same level, but then maybe the relevant package/class hierarchy should be reconsidered). So it'll be something like `[...].graphs.SelectionPolicy` and `[...].graphs.EdgeSelectionPolicy` which IMO makes sense because `Edge` there is not an alternative to `Graph` but instead just an amendment on top of it.

You note yourself: "these models simply compute the lists of indices, and don't actually perform selection or inspection". So then "selection", "inspection", and "hit test" don't make sense here. A graph has a full set of possible indices, and by finding out a subset of those indices all you're doing is simply filtering - so why not use the word `Filter`?

Finally, I myself don't like the word `Policy` that's used throughout. It has some strong declarative connotations that aren't apt, especially since we want to support `CustomJS` or something similarly flexible and dynamic. Given the previous paragraph, it's possible to remove the word `Policy` altogether. As an alternative, something less declarative and with fewer connotations could be used, e.g. `Rule`.

As an example of all naming notes together, omitting the initial `[...]`:
```
graphs/index_filters.py:
- NodeFilter
- AdjacentNodes(NodeFilter)
- EdgeFilter
- AdjacentEdges(EdgeFilter)

graphs.py or graphs/__init__.py:
- IndexFilter
```

Note that it's `NodeFilter` and not `NodeIndexFilter`. Personally, I prefer it that way because that class is already in the context of `index_filters` - we already know _what_ it is, so `Index` is superfluous. And we still need `Filter` in there because it's a very simple abstract class, there are no words that define its behavior because there is no behavior, so we have to add something.

Addressing my anticipation of the "but such short names will then be ambiguous after imports" argument. That's only the case when you import classes by themselves. Experience of other languages shows that, given that it's very much feasible in Python, importing classes themselves is not a good idea. Apart from name clashes, such imports also grow with each new name you need, requiring you to change your sources in multiple places. Instead, in this particular case something like `import [...].graphs.index_filters as gi_flt` could be used, with subsequent `gi_flt.AdjacentNodes`. It can be as verbose or as terse as one wants, while significantly reducing the imports block size, the amount of changes throughout the lifetime of a project, and the chance of name clashing. And as a bonus, it also gives a library's maintainers to choose whatever names they want for their classes, without inventing mile-long names reminiscent of the Hungarian notation.

It turned out quite a bit longer than I anticipated - sorry about that. But I also can't really make it shorter without removing potentially useful bits.

**Reply by bryevdv** [reply]
@p-himik Thanks for the the thorough response, comments inline:

>  I still feel that there might be a mechanism that would enable a uniform and flexible interface to have arbitrary filters without having to write too much boilerplate code. Let me know, and if so, I'll try to come up with something in a few days.

I would definitely be interested in this! Perhaps a good initial scope is just as you suggest:

* comparison for a single node or edge attr
* comparison for computed sum of edge weights on paths

Regarding adjacent node "hops", I am fairly keen to still include it just because we have to handle 0 and 1 hops and it's trivial to generalize. I also *can* easily imagine being asked for it, e.g. to be able to highlight "regions" around nodes. So there is little incentive not to do it up front in my mind. 

> "these models simply compute the lists of indices, and don't actually perform selection or inspection"

This was very poorly expressed on my part. By the above, I meant only that they don't have to be responsible for the specific narrow action of *setting the actual `.selection` property*. Yes, could "just compute indices", but the indices they compute are always and only:

* derived from hit-testing
* to be set as a selection

That context is really inseparable form their purpose, does that alter your appraisal here at all? That said, I do think "filter" is a good conceptual frame to apply here in some capacity. 

> I myself don't like the word Policy that's used throughout. 

Well, I picked this up doing objective-C coding many years ago, where any kind of little thunk to control a local behavior was referred to as a policy. It definitely does not connote declarative *to me*. In any event we have used it alot in this codebase over several years, and at this point I am by default inclined to favor self-consistency over anything else. There is also the matter that the current property is already called `selection_policy` and I'd like to afford some migration path.  Given that, and the above, I would use "Policy" for the top-level class, and instances of those would be comprise of collections of "filters". How does that seem?

> but such short names will then be ambiguous after imports

These are models, and `bokeh.models` is ultimately a flat namespace. This is due to both, but also due to technical limitations with how models are serialized across Python<-->JS. Changing that would be a *much* bigger scope. Barring someone signing up for a much bigger undertaking, I think we have to live with explicit and unique name for models.

**Reply by p-himik** [reply]
> That context is really inseparable form their purpose

Not necessarily. With the right architecture, it should be trivial to compose the pieces of functionality that actually define how you want to filter for indices. For example, "highlight all nodes one hop away from all the neighbors of the current node connected to it with edges with weight < 0.5" - there are two separate filters here, combined, just as we already have in `CDSView` filters. So the overall context is still the same, but individual filters are used within the context of something that combines them - they don't care about it being a selection.

And that actually makes me think that selection, highlighting, _and_ data filtering are all conceptually the same. I imagine combining them all by using primitive models would not only make a few things simpler, but also enable algorithmic selection - because you'd be able to use filters that are used with `CDSView` for selection and highlighting.

> I would use "Policy" for the top-level class, and instances of those would be comprise of collections of "filters". How does that seem?

Sounds alright, but I'd have to see what "comprised of collections of filters" means exactly. Because you just described it in a very similar way to `CDSView` yourself, and I definitely like that, given the previous paragraph.

> `bokeh.models` is ultimately a flat namespace

Right. And I imagine you don't want to stop putting everything into a single package just because it has been done that way forever? If it were up to me, I'd probably try to [gradually] steer away from the catch-all `bokeh.models` and into `bokeh.models.foo`, `bokeh.models.bar`, etc everywhere - documentation, examples, recommendations.

> due to technical limitations with how models are serialized across Python<-->JS. Changing that would be a much bigger scope.

I see. On the bright side, such an undertaking would make Bokeh and BokehJS more similar than they are now (assuming Python and JS packages are mapped by matching names and not by some explicit map of strings) and would make BokehJS significantly more approachable. Perhaps an idea for 4.0? :)

**Reply by p-himik** [reply]
> I would definitely be interested in this! Perhaps a good initial scope is just as you suggest:
> - comparison for a single node or edge attr
> - comparison for computed sum of edge weights on paths

This ended up being a bit more complicated than I'd hope, but I think there are two main approaches.

One is a fully declarative and flexible solution, but it requires significant efforts for both implementing it as a developer and getting acquainted with it as a user. Namely, people have suggested the Datalog query language - it (or things analogous to it) is used in many areas where graph querying is involved. With that being said, it would be nice to have it available for Bokeh but I don't think it belongs in Bokeh codebase.

The other approach that is simple enough and that should cover a reasonable amount, if not the majority, of use-cases is to represent node/edge selection as a sequence of operations where each operation is either an expansion of the working set or its filtering. Initially, the working set consists of all the selected/highlighted nodes/edges. To address the initial scope you provided:
Comparison for a single node or edge attr:
```python
[ExpandToNeighbors(),  # default hops=1
 FilterEdgesByAttr(attr='weight', value=0.5, op='lt'),  # also removes the nodes that are no longer connected within the expanded working set subgraph
 FilterNodesByAttr(attr='kind', value='cat'),  # default op='eq'
 OnlyNodes()]  # optionally, leave only nodes to avoid highlighting the edges
```
Comparison for computed sum of edge weights on paths:
```python
[ExpandToAll(),  # the originally selected/highlighted set should still be available somewhere for the next step
 FilterNodesByPathLength(attr='weight')]  # this seems common enough to deserve its own class
```
Of course, potentially one can interleave expansion with filtering. It's unclear how it must work with `FilterNodesByPathLength` - I'd say the initial implementation should only consider paths between each node in the initially selected/highlighted set and each node that's in the expanded working set but outside of the initial set. Same with other filters - unclear whether they should apply to all the nodes/edges already in the working set or only nodes/edges that were added during the previous expansion step. I think any approach works here as a default, and then a new optional argument could be introduced if needed.

With the above, it's clear that this approach doesn't allow for many scenarios but it should be extensible enough. And, as I mentioned, that Datalog querying can be at some point implemented as probably a separate library - which should be possible as long as built-in Bokeh functionality remains extensible.

**Reply by bryevdv** [reply]
Thanks for the continued discussion @p-himik thoughts below

---
(Brief aside) 

> Right. And I imagine you don't want to stop putting everything into a single package just because it has been done that way forever? 

Not really. We used to expose all submodules early on, but this caused pain on several occasions when we needed to re-arrange things. In fact the earnest move to expose *only* `bokeh.models` as a documentation practice is fairly recent, specifically to insulate against that risk. I think namespaces are good (of course) but the actual situation here is that the fine-grained submodule structure (for organizing development) is not a good namespace structure (for UX and documentation). The right solution would be intentional, explicit, coarser grained namespace modules e.g. a `bokeh.models.widgets` might collect all the widgets. I'd probably get behind a subdivision into 4-6 sub-namespaces, much more than that at the subdivisions themselves start imposing their own cognitive burden. 

As a separate aside, I just really hate reading code like `gi_flt.AdjacentNodes` inline. I vastly prefer importing individual names once at the top, so that they may be used cleanly and unqualified throughout the actual code. So I'm probably going to tilt towards unique names even if explicitly namespaces are introduced. 

---

Anyway, back to the task at hand. I think your proposal above is essentially congruent to my original proposal at the top, except with a bit of a different POV/emphasis on the available filters. But I do have one question. It's possible to "start" a selection [1] from an edge or a node. What does `ExpandToNeighbors` mean, exactly in those two cases? Or put another way: if I hover over a node, I might want to highlight:

* nothing else
* only adjacent node
* only immediate edges
* both

I think it would be instructive to see how all four of these would be spelled using your approach to filter nomenclature. 


[1] BTW more sloppiness here I guess. Any time I say "selection" I really mean "selection or inspection" since they are identical in execution. This could/would also include "programmatic selection" or "programmatic graph filtering" in case we wanted to expose some notion of that to users as well.

**Reply by p-himik** [reply]
> The right solution would be intentional, explicit, coarser grained namespace modules e.g. a `bokeh.models.widgets` might collect all the widgets.

That makes sense!

> I vastly prefer importing individual names once at the top, so that they may be used cleanly and unqualified throughout the actual code.

But what about `import pandas as pd` and `import numpy as np`? ;)

The two approaches can co-exist - either, as you say, by keeping the names unique in Bokeh code base (if so, there will probably have to be a test for it eventually) _or_ at the user end by using imports like `import graph.index_filters.AdjacentNodes as IndexFilterAdjacentNodes`. Either is fine, it's just that personally I wouldn't want to keep track of names while making sure they're unique and different enough to avoid any confusion.

> What does `ExpandToNeighbors` mean, exactly in those two cases?

First of all I should mention that my proposal above is not something incredibly robust that one could start working on - more design work is needed while paying close attention to all the desired supported scenarios. While I try to keep in mind things I personally find reasonable or potentially desirable, I never actually work with graphs so I could be completely wrong in that domain.

With that being said, it could probably be renamed to `ExpandToNodes`. When selecting nodes it would expand to other nodes and when selecting edges it would expand first to the nodes that are connected by those edges, and then to the nodes beyond those if `hops > 1`. It should be possible to handle all four scenarios you mention by classes like `OnlyNodes`.

**Reply by bryevdv** [reply]
I guess most of the answer above comes down to e.g. adding `OnlyNodes()` or `OnlyEdges()` to the filters. I guess that leaves users open the possibility to do non-sensical things like add both.  

I'm more unsure about e.g. `FilterEdgesByAttr` In my mind it would be common to want to "look at all nodes a distance of 10 away" but then this policy needs to filter *nodes, but according to edges* Would you propose a separate `FilterNodesbByEdgeAttr` ? Or have `FilterEdgesByAttr` also implicitly cull nodes? That seems like it might become hard to reason about.

**Reply by bryevdv** [reply]
In any event I will try to reformulate a proposal that enumerates common cases, with this different filtering POV in mind.

**Reply by p-himik** [reply]
>  that leaves users open the possibility to do non-sensical things like add both.

Yep, an intersection of all nodes and all edges is an empty set. Nothing would be highlighted. Or it would be an error - depends on implementation.

> In my mind it would be common to want to "look at all nodes a distance of 10 away" but then this policy needs to filter nodes, but according to edges

I'm afraid I don't follow. Looking at nodes 10 hops away is just expanding the working set 10 times, or by 10 hops - there's no filtering involved.
`FilterEdgesByAttr` removes edges _and_ all connected nodes. In other words, if filters out paths in graph - my whole approach is about paths, not nodes or edges specifically. It makes things with divergent paths like "selected red edges but forget about their nodes and select nodes connected by blue edges, but forget the blue edges themselves" impossible without either inventing some nasty classes specifically for that or turning that flat path-expanding-and-filtering list into a proper tree. Which is not that bad - it's just nested lists - but it might get confusing.

**Reply by bryevdv** [reply]
> I'm afraid I don't follow. Looking at nodes 10 hops away is just expanding the working set 10 times, or by 10 hops - there's no filtering involved.

Sorry, by distance I did not mean hops, I meant a path distance according so some edge attr that provides "distance".  In any case like I said I will try to come up with ~10 scenarios that seems like they would be the most common, and hopefully we can start from there and see how things would be spelled explicitly.

**Reply by p-himik** [reply]
Ah, that is handled by my second code block in the initial proposal, with `FilterNodesByPathLength`. Seems like I forgot to add a number there, so it should be `FilterNodesByPathLength(attr='weight', value=10, op='le')`.

Indeed, looking forward!
